### ISSUE TITLE
Reconstruction of sparse model from known camera poses gives few 3D points

### ISSUE THREAD

**Issue by char-lie** [colmap/colmap#497]
I use a calibration object to know positions and directions of cameras in my static rig.
Then, I try to follow the instruction [Reconstruct sparse/dense model from known camera poses](https://colmap.github.io/faq.html#reconstruct-sparse-dense-model-from-known-camera-poses).
My OS is Xubuntu 18.04.

First, I create a sparse model with cameras' positions

- I convert binary sparse model files to text

```bash
colmap model_converter \
    --input_path calibration/sparse/ \
    --output_path calibration/cameras \
    --output_type TXT
```

- I make `points3D.txt` file blank, because the instruction says `The points3D.txt file should be empty`
- In `images.txt` I clean lines that contain information about points; though, it was not clear for me: FAQ says `every other line in the images.txt should also be empty` without specifying what `every other line` means (should it be clarified in FAQ?)
- I leave `cameras.txt` file untouched

Then, I extract and match features without any problems

```bash
colmap feature_extractor \
    --image_path model_images/ \
    --database_path output/db.db \
    --SiftExtraction.max_image_size 4096 \
    --SiftExtraction.max_num_features 16384 \
    --SiftExtraction.estimate_affine_shape 1 \
    --SiftExtraction.domain_size_pooling 1 \
    
colmap exhaustive_matcher \
    --database_path output/db.db \
    --SiftMatching.guided_matching 1
```

Note that `model_images` are not the same as were used for camera calibration.

Finally, triangulation fails

```bash
colmap point_triangulator \
    --database_path output/db.db \
    --image_path model_images/ \
    --input_path calibration/cameras/ \
    --output_path output/triangulated
```

I caught an error caused by different identifiers of images between calibration sparse model and new model that I want to reconstruct

```bash
==============================================================================
Loading database
==============================================================================

Loading cameras... 24 in 0.000s
Loading matches... 32 in 0.000s
Loading images... 24 in 0.004s (connected 20)
Building correspondence graph... in 0.001s (ignored 0)

Elapsed time: 0.000 [minutes]

F1018 10:55:52.417294   460 reconstruction.cc:81] Check failed: existing_image.Name() == image.second.Name() (17.JPG vs. 16.JPG) 
*** Check failure stack trace: ***
    @     0x7f941eb240cd  google::LogMessage::Fail()
    @     0x7f941eb25f33  google::LogMessage::SendToLog()
    @     0x7f941eb23c28  google::LogMessage::Flush()
    @     0x7f941eb26999  google::LogMessageFatal::~LogMessageFatal()
    @     0x560163359977  (unknown)
    @     0x5601634be763  (unknown)
    @     0x5601632ba0d4  (unknown)
    @     0x5601632ac545  (unknown)
    @     0x7f941a400b97  __libc_start_main
    @     0x5601632b622a  (unknown)
Aborted (core dumped)
```

I've changed ids by hand and got very few points triangulated

```bash
==============================================================================
Loading database                                                                                                                  
==============================================================================                                                           
                                                                                                           
Loading cameras... 24 in 0.000s                                                                            
Loading matches... 26 in 0.000s                                                                            
Loading images... 24 in 0.008s (connected 19)                                                              
Building correspondence graph... in 0.002s (ignored 0)                                                     
                                                                                                           
Elapsed time: 0.000 [minutes]                                                                              
                                                                                                           
                                                                                                           
==============================================================================                                         
Triangulating image #16                                                       
==============================================================================
                                                                              
  => Image has 0 / 526 points                                                 
  => Triangulated 141 points                                                  
                                                                              
==============================================================================
Triangulating image #13                                                       
==============================================================================
                                                                              
  => Image has 0 / 298 points                                                 
  => Triangulated 54 points                                                   
                                                                              
==============================================================================
Triangulating image #15                                                       
==============================================================================
                                                                                                           
  => Image has 0 / 309 points                                                                              
  => Triangulated 26 points                                                                                
                                                                                                           
==============================================================================                             
Triangulating image #14                                                                                    
==============================================================================                             
                                                                                                           
  => Image has 2 / 168 points                                                 
  => Triangulated 5 points                                                    
                                                                              
==============================================================================
Triangulating image #12                                                       
==============================================================================
                                                                              
  => Image has 0 / 835 points                                                 
  => Triangulated 209 points                                                  
                                                                              
==============================================================================                                                    
Triangulating image #11                                                                                                                  
==============================================================================                             
                                                                                                           
  => Image has 0 / 0 points                                                                                
  => Triangulated 0 points                                                                                 

==============================================================================
Triangulating image #10
==============================================================================

  => Image has 0 / 0 points
  => Triangulated 0 points

==============================================================================
Triangulating image #9
==============================================================================

  => Image has 0 / 394 points
  => Triangulated 118 points

==============================================================================
Triangulating image #18
==============================================================================

  => Image has 13 / 118 points
  => Triangulated 19 points

==============================================================================
Triangulating image #4
==============================================================================

  => Image has 102 / 859 points
  => Triangulated 75 points

==============================================================================
Triangulating image #20
==============================================================================

  => Image has 0 / 0 points
  => Triangulated 0 points

==============================================================================
Triangulating image #1
==============================================================================

  => Image has 0 / 233 points
  => Triangulated 0 points

==============================================================================
Triangulating image #22
==============================================================================

  => Image has 0 / 0 points
  => Triangulated 0 points

==============================================================================
Triangulating image #3
==============================================================================

  => Image has 113 / 933 points
  => Triangulated 79 points

==============================================================================
Triangulating image #21
==============================================================================

  => Image has 0 / 146 points
  => Triangulated 20 points

==============================================================================
Triangulating image #5
==============================================================================

  => Image has 90 / 669 points
  => Triangulated 74 points

==============================================================================
Triangulating image #23
==============================================================================

  => Image has 0 / 99 points
  => Triangulated 12 points

==============================================================================
Triangulating image #6
==============================================================================

  => Image has 173 / 971 points
  => Triangulated 88 points

==============================================================================
Triangulating image #24
==============================================================================

  => Image has 0 / 233 points
  => Triangulated 0 points

==============================================================================
Triangulating image #8
==============================================================================

  => Image has 175 / 780 points
  => Triangulated 14 points

==============================================================================
Triangulating image #19
==============================================================================

  => Image has 104 / 707 points
  => Triangulated 89 points

==============================================================================
Triangulating image #2
==============================================================================

  => Image has 25 / 463 points
  => Triangulated 9 points

==============================================================================
Triangulating image #17
==============================================================================

  => Image has 0 / 0 points
  => Triangulated 0 points

==============================================================================
Triangulating image #7
==============================================================================

  => Image has 9 / 292 points
  => Triangulated 30 points
  => Merged observations: 0
  => Completed observations: 15
```

Surface I want to reconstruct has not a lot of SIFT matches, so the procedure provides bad sparse point cloud, and patch match processes only 16 images of 24.

If I avoid triangulation and use result of sparse reconstruction of calibration object in patch match procedure, I have all 24 images processed and dense reconstruction result is much more complete.

```bash
colmap image_undistorter \
    --input_path calibration/sparse/ \
    --image_path model_images/ \
    --output_path output/
colmap patch_match_stereo \
    --PatchMatchStereo.geom_consistency 1 \
    --workspace_path output/
```

Am I doing all right?
I have an assumption that my calibration object is similar to scanned objects in that sense that they have the same neighboring images and direct reconstruction of dense point cloud works fine.
Though, it's strange: I use calibration object just because I cannot match features of another images well, and, according to FAQ, I should use features matching to use calibration object.

**Comment by ahojnnes** [comment]
The image_id in the database must match the image_id in the images.txt. Also make sure that the images.txt contains exactly the same image observations as listed in the keypoints table in the database. Note that you simply set the point3D_id of each observation to -1.

**Comment by char-lie** [comment]
@ahojnnes, got it. Thank you!

**Comment by pilafniev** [comment]
Hello!
I have been trying to do exactly this but I haven't been able to. I get the same error as char-lie. Hopefully one of you is able to help me.
This is the steps I have taken.
1) Make sparse model of calibration item
2) Convert model to TXT format just like mentioned above.
3) Copy images.txt, cameras.txt, and 3Dpoints.txt to my new model folder (DATASET_PATH)
4) Leave cameras.txt untouched, erase all data in 3Dpoint.txt, and erase every other line in images.txt so only 
`#   IMAGE_ID, QW, QX, QY, QZ, TX, TY, TZ, CAMERA_ID, NAME `
lines are left, with a blank line in between as such:
```
# Image list with two lines of data per image:
#   IMAGE_ID, QW, QX, QY, QZ, TX, TY, TZ, CAMERA_ID, NAME
#   POINTS2D[] as (X, Y, POINT3D_ID)
# Number of images: 100, mean observations per image: 952.95
100 0.989123 0.0712152 -0.114599 -0.0585765 -0.0102577 -1.16044 3.1816 100 image99.jpg

99 0.984757 -0.143058 -0.0987996 -0.0052357 0.19831 -2.07697 4.13605 98 image97.jpg

98 0.990242 -0.055276 -0.120138 -0.0439669 -0.131827 -1.82626 3.39713 99 image98.jpg

97 0.949651 0.308238 -0.00655007 0.0557715 0.10588 -1.0627 2.41078 97 image96.jpg

96 0.997251 0.0739932 -0.00246529 -0.0029146 -0.0276835 -1.16819 3.17813 96 image95.jpg
```

5) Run feature extractor and exhaustive matcher in new model folder which creates its new its own database.
```
DATASET_PATH=$(pwd)
colmap feature_extractor \
--database_path $DATASET_PATH/database.db \
--image_path $DATASET_PATH/images \
--SiftExtraction.max_image_size 3280 \
--SiftExtraction.estimate_affine_shape 1 \
--SiftExtraction.domain_size_pooling 1
colmap exhaustive_matcher \
--database_path $DATASET_PATH/database.db \
--SiftMatching.multiple_models 1 \
--SiftMatching.guided_matching 1 \
--ExhaustiveMatching.block_size 20
mkdir $DATASET_PATH/sparse
```
6) Finally run mapper command
```
colmap mapper \
--database_path $DATASET_PATH/database.db \
--image_path $DATASET_PATH/images \
--image_input $DATASET_PATH
--output_path $DATASET_PATH/sparse
```
But I end up getting this error 
```
Loading cameras... 100 in 0.000s
Loading matches... 1868 in 0.010s
Loading images... 100 in 0.017s (connected 100)
Building correspondence graph... in 0.112s (ignored 0)

Elapsed time: 0.002 [minutes]

F1109 16:23:19.704193 37228544 reconstruction.cc:81] Check failed: existing_image.Name() == image.second.Name() (image98.jpg vs. image95.jpg) 
*** Check failure stack trace: ***
    @        0x10bb19efd  google::LogMessage::SendToLog()
    @        0x10bb1a677  google::LogMessage::Flush()
    @        0x10bb1dbf7  google::LogMessageFatal::~LogMessageFatal()
    @        0x10bb1af89  google::LogMessageFatal::~LogMessageFatal()
    @        0x10b06bb05  colmap::Reconstruction::Load()
    @        0x10b28e324  colmap::IncrementalMapper::BeginReconstruction()
    @        0x10b0b9dfb  colmap::IncrementalMapperController::Reconstruct()
    @        0x10b0b9634  colmap::IncrementalMapperController::Run()
    @        0x10b3a9d3d  colmap::Thread::RunFunc()
    @        0x10b3abb3e  _ZNSt3__114__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEMN6colmap6ThreadEFvvEPS8_EEEEEPvSD_
    @     0x7fff6c76b6c1  _pthread_body
    @     0x7fff6c76b56d  _pthread_start
Abort trap: 6

```
I have tried this with a different set of images (same quantity and name) as well as with a copy of the original calibration model folder, (same pictures) but have not had success.

I would greatly appreciate it I could get some feedback on what I'm doing incorrectly.
Also wondering : 
1) Is it correct to create a new data base with the new set of pictures or should I be using the database from the calibration model and modifying it somehow?
2) Am I modifying images.txt correctly? 
3) Why would I be getting an error even when performing these steps on the same calibration model images?

Thanks in Advance !

**Comment by char-lie** [comment]
@pilafniev, have you opened a database with features fetched on step `6` (resulting database `$DATASET_PATH/database.db` after `mapper` command run) and checked ids of all images?
You can do this by running `colmap gui`, there you create a new project and `Open` an existing database

![Steps to open database](https://user-images.githubusercontent.com/5962274/48301379-bb949f00-e4f5-11e8-9b04-3dae6ac5a98b.png)

Then, you select directory with images in usual way and click `Save`.

Finally, you click `Database management` button (located to the right from `Feature extraction` and `Feature matching`) and see names of images and their identifiers

![image](https://user-images.githubusercontent.com/5962274/48301401-3cec3180-e4f6-11e8-90b7-1d8e6231812d.png)

In this case ids are the same as names, but they may differ like in your case.
For example, in your `images.txt` file you have an image `image97`.
Its id is `99` (the first number in the row).

```
99 0.984757 -0.143058 -0.0987996 -0.0052357 0.19831 -2.07697 4.13605 98 image97.jpg
```

If you had few images (in my case `7`), you could edit the `images.txt` file manually and specify the same ids like in `database.db`.
Though, if you have a lot of them, you should probably write a script that loads the database using SQLite driver and change identifiers automatically.

Though, for the start I'd recommend to use 5-10 images to reconstruct a part of the object and check whether it works.

**Comment by pilafniev** [comment]
Thanks for the quick response Char-lie!
I did have some mismatched image id's.. which were causing the error. While on the subject, what other tools or advise would you recommend for modeling differences between objects? I want to model 2 very similar objects with a very small variation between them, but when transitioning from one model to the other I would like for only the differences to be noticeable in the transition and keep everything else static. I'd appreciate any feedback on the subject, thanks again!

**Comment by char-lie** [comment]
@pilafniev, let's chat under the gist https://gist.github.com/char-lie/61d74885eaf4d62fcf1b959f7f7947d5 to not add noise to this topic. Add a comment there for me to know that you're there.

**Comment by silence401** [comment]
I just have the camera poses but i don't have camera intrinsic.So should I leave the cameras.txt blank or the cameras.txt must have something.I got some error when the cameras.txt is blank.
==============================================================================
Loading model
==============================================================================

terminate called after throwing an instance of 'std::invalid_argument'
  what():  stold
*** Aborted at 1562074487 (unix time) try "date -d @1562074487" if you are using GNU date ***
PC: @     0x7f2f3b900e97 gsignal
*** SIGABRT (@0x3e800007748) received by PID 30536 (TID 0x7f2f405b2bc0) from PID 30536; stack trace: ***
    @     0x7f2f3eb66890 (unknown)
    @     0x7f2f3b900e97 gsignal
    @     0x7f2f3b902801 abort
    @     0x7f2f3c524957 (unknown)
    @     0x7f2f3c52aab6 (unknown)
    @     0x7f2f3c52aaf1 std::terminate()
    @     0x7f2f3c52ad24 __cxa_throw
    @     0x7f2f3c5267e3 (unknown)
    @     0x560288c5c1b2 (unknown)
    @     0x560288c62a44 (unknown)
    @     0x560288c67d7b (unknown)
    @     0x560288c68409 (unknown)
    @     0x560288bc2cc4 (unknown)
    @     0x560288bb447d (unknown)
    @     0x7f2f3b8e3b97 __libc_start_main
    @     0x560288bbe1ca (unknown)

**Comment by HareshMiriyala** [comment]
i had the same issue, and figured out that the images need to saved as 001.png and not 1.png 

If you're using python, you can use the zfill inbuilt function to solve this

**Comment by chenhao2015** [comment]
> 
Hi, @ahojnnes , @char-lie 

May I inquire what does it mean  "make sure that the images.txt contains exactly the same image observations as listed in the keypoints table in the database." I also meet the problem "Reconstruction of sparse model from known camera poses gives few 3D points". I only got 50 points.

My commands:

colmap feature_extractor --image_path /home/haochen/NeRF/report/images --database_path /home/haochen/NeRF/report/database.db  --SiftExtraction.max_image_size 4096 --SiftExtraction.max_num_features 16384 --SiftExtraction.estimate_affine_shape 1 --SiftExtraction.domain_size_pooling 1 --ImageReader.camera_model SIMPLE_PINHOLE --ImageReader.single_camera 1

colmap exhaustive_matcher --database_path /home/haochen/NeRF/report/database.db --SiftMatching.guided_matching 1

colmap point_triangulator --database_path /home/haochen/NeRF/report/database.db --image_path /home/haochen/NeRF/report/images --input_path /home/haochen/NeRF/report/sparse/created/model --output_path /home/haochen/NeRF/report/sparse/model


Please see the following logs of point_triangulator .

==============================================================================
Loading model
==============================================================================


==============================================================================
Loading database
==============================================================================

Loading cameras... 1 in 0.000s
Loading matches... 136 in 0.001s
Loading images... 17 in 0.001s (connected 17)
Building correspondence graph... in 0.008s (ignored 0)

Elapsed time: 0.000 [minutes]


==============================================================================
Triangulating image #17 (0)
==============================================================================

  => Image sees 0 / 1478 points
  => Triangulated 0 points

==============================================================================
Triangulating image #15 (1)
==============================================================================

  => Image sees 665 / 1765 points
  => Triangulated 877 points

==============================================================================
Triangulating image #9 (2)
==============================================================================

  => Image sees 1235 / 1720 points
  => Triangulated 264 points

==============================================================================
Triangulating image #8 (3)
==============================================================================

  => Image sees 1022 / 1684 points
  => Triangulated 110 points

==============================================================================
Triangulating image #12 (4)
==============================================================================

  => Image sees 1171 / 1642 points
  => Triangulated 79 points

==============================================================================
Triangulating image #11 (5)
==============================================================================

  => Image sees 246 / 1579 points
  => Triangulated 26 points

==============================================================================
Triangulating image #13 (6)
==============================================================================

  => Image sees 845 / 1589 points
  => Triangulated 29 points

==============================================================================
Triangulating image #14 (7)
==============================================================================

  => Image sees 1410 / 1799 points
  => Triangulated 153 points

==============================================================================
Triangulating image #10 (8)
==============================================================================

  => Image sees 1529 / 1803 points
  => Triangulated 127 points

==============================================================================
Triangulating image #16 (9)
==============================================================================

  => Image sees 1419 / 1811 points
  => Triangulated 166 points

==============================================================================
Triangulating image #4 (10)
==============================================================================

  => Image sees 1311 / 1518 points
  => Triangulated 38 points

==============================================================================
Triangulating image #7 (11)
==============================================================================

  => Image sees 427 / 1634 points
  => Triangulated 52 points

==============================================================================
Triangulating image #2 (12)
==============================================================================

  => Image sees 0 / 1529 points
  => Triangulated 0 points

==============================================================================
Triangulating image #3 (13)
==============================================================================

  => Image sees 1521 / 1646 points
  => Triangulated 18 points

==============================================================================
Triangulating image #1 (14)
==============================================================================

  => Image sees 1451 / 1595 points
  => Triangulated 40 points

==============================================================================
Triangulating image #6 (15)
==============================================================================

  => Image sees 1416 / 1559 points
  => Triangulated 27 points

==============================================================================
Triangulating image #5 (16)
==============================================================================

  => Image sees 1439 / 1554 points
  => Triangulated 31 points

==============================================================================
Retriangulation
==============================================================================

  => Completed observations: 0
  => Merged observations: 0

==============================================================================
Bundle adjustment
==============================================================================

iter      cost      cost_change  |gradient|   |step|    tr_ratio  tr_radius  ls_iter  iter_time  total_time
   0  3.794887e+08    0.00e+00    6.95e+02   0.00e+00   0.00e+00  1.00e+04        0    8.84e-03    1.84e-02
   1  3.100615e+08    6.94e+07    1.73e+02   4.51e+05   8.19e-01  1.35e+04        0    1.39e-02    3.24e-02
   2  2.978513e+08    1.22e+07    4.32e+01   6.78e+05   8.06e-01  1.75e+04        0    1.29e-02    4.53e-02
   3  2.954552e+08    2.40e+06    1.07e+01   9.30e+05   8.13e-01  2.33e+04        0    1.30e-02    5.82e-02
   4  2.949659e+08    4.89e+05    2.54e+00   1.07e+06   8.35e-01  3.33e+04        0    1.27e-02    7.10e-02
   5  2.948598e+08    1.06e+05    5.74e-01   1.10e+06   8.61e-01  5.34e+04        0    1.27e-02    8.37e-02


Bundle adjustment report
------------------------
    Residuals : 38738
   Parameters : 20979
   Iterations : 6
         Time : 0.0844555 [s]
 Initial cost : 98.9762 [px]
   Final cost : 87.2447 [px]
  Termination : Convergence

  => Completed observations: 0
  => Merged observations: 0
  => Filtered observations: 19269
  => Changed observations: 0.994837

==============================================================================
Bundle adjustment
==============================================================================

iter      cost      cost_change  |gradient|   |step|    tr_ratio  tr_radius  ls_iter  iter_time  total_time
   0  2.992714e+02    0.00e+00    9.09e-12   0.00e+00   0.00e+00  1.00e+04        0    5.07e-05    1.32e-04


Bundle adjustment report
------------------------
    Residuals : 200
   Parameters : 150
   Iterations : 1
         Time : 0.000152124 [s]
 Initial cost : 1.22326 [px]
   Final cost : 1.22326 [px]
  Termination : Convergence

  => Completed observations: 0
  => Merged observations: 0
  => Filtered observations: 0
  => Changed observations: 0.000000

==============================================================================
Extracting colors
==============================================================================

**Comment by meyerls** [comment]
Maybe a little bit late but i found out that if you use 
```    
--SiftExtraction.estimate_affine_shape 1 \
 --SiftExtraction.domain_size_pooling 1 \
```
the order of the image will change and does not match with the order of your original database.db / images.bin file.

If you remove this part the sift feature extraction time will roughly be the same and the images will be loaded in the identical order (mostly).

**Comment by JayNyce** [comment]
I am running into this error during colmap patch_match_stereo
Reading workspace...
Reading configuration...
Configuration has 420 problems...

==============================================================================
Processing view 1 / 420 for frame_0000336000.png
==============================================================================

Reading inputs...
F1025 13:11:42.196605 19424 image.cc:56] Check failed: width_ == bitmap_.Width() (1199 vs. 1200)
*** Check failure stack trace: ***
    @   00007FFB66FD5FFB  public: void __cdecl google::LogMessage::Flush(void) __ptr64
    @   00007FFB66FD4B49  public: __cdecl google::LogMessageFatal::~LogMessageFatal(void) __ptr64
    @   00007FF7A21C0D71  (unknown)
    @   00007FF7A22C9E59  (unknown)
    @   00007FF7A1FB37FF  (unknown)
    @   00007FF7A203D734  (unknown)
    @   00007FF7A203B938  (unknown)
    @   00007FFB7D099363  _recalloc
    @   00007FFB7E12257D  BaseThreadInitThunk
    @   00007FFB7F48AA78  RtlUserThreadStart 

Anyone with any ideas how to solve it

**Comment by MyLovelyAxe** [comment]
Hello, I am trying to use known camera poses for sparse reconstruction. I use pycolmap API and .bin files by the way, instead of .txt files mentioned [here in FAQ of colmap document](https://colmap.github.io/faq.html#reconstruct-sparse-dense-model-from-known-camera-poses). These are what I did:

1) Generate cameras.bin with the same content in database.db
2) Generate images.bin as introduced in document, leaving the xys and points3D_id blank with `write_next_bytes(fid, 0, "Q")` from `colmap/scripts/python/read_write_model.py`
3) Generate a blank points3D.bin just `write_next_bytes(fid, 0, "Q")`

After `pycolmap.extract_features` and `pycolmap.match_sequential`, I use `pycolmap.incremental_mapping`, but the issue is that the points3D.bin remains blank, without any content, as well as points3D_id in images.bin. I can confirm that feature extraction and feature matching works fine since in both COLMAP GUI database management and reading-out by `colmap/scripts/python/read_write_model.py` of xys in images.bin give reasonable results. I am confused why only points3D-related part (i.e. points3D.bin and points3d_id in images.bin) doesn't update.

I noticed from this reply:
> The image_id in the database must match the image_id in the images.txt. Also make sure that the images.txt contains exactly the same image observations as listed in the keypoints table in the database. Note that you simply set the point3D_id of each observation to -1.
I am not sure what does it mean `Note that you simply set the point3D_id of each observation to -1.` I understand label -1 means no matching observation for 3D point, but before mapping, how do I know how many 3D points it will have and specify the amount of -1?

Besides, should I stick to use .txt instead of .bin? Or use `point_triangulator` instead of `incremental mapping`? Thank you very much for any advices.
